cirq/circuits/circuit_operation.py

```diff
@@ -188,53 +262,97 @@ def _json_dict_(self):
             'qubit_map': sorted(self.qubit_map.items()),
             'measurement_key_map': self.measurement_key_map,
             'param_resolver': self.param_resolver,
+            'repetition_ids': self.repetition_ids,
         }
 
     @classmethod
     def _from_json_dict_(
-        cls, circuit, repetitions, qubit_map, measurement_key_map, param_resolver, **kwargs
+        cls,
+        circuit,
+        repetitions,
+        qubit_map,
+        measurement_key_map,
+        param_resolver,
+        repetition_ids,
+        **kwargs,
     ):
         return (
             cls(circuit)
             .with_qubit_mapping(dict(qubit_map))
             .with_measurement_key_mapping(measurement_key_map)
             .with_params(param_resolver)
-            .repeat(repetitions)
+            .repeat(repetitions, repetition_ids)
         )
 
     # Methods for constructing a similar object with one field modified.
 
     def repeat(
-        self,
-        repetitions: INT_TYPE,
+        self, repetitions: INT_TYPE, repetition_ids: Optional[List[str]] = None
```

Comment:
What do you think about making both args optional? I think it'd be nice for `repeat(['a', 'b', 'c'])` to work without specifying the number of IDs. (Requires addressing `repetitions` vs `repetitions * self.repetitions` first)

Reply:
That was what I had originally implemented but because of multiple optionals, the default-ing logic was becoming too complicated and I felt it was making the API messier. But if we are making the override-all-IDs case more explicit, I should be able to add this in as well.

Reply:
Made `repetitions` `Optional` as well. The actual conditions of what defaults to what are a little difficult to describe but added lot of docstrings to compensate. Also, even though it's verbose to explain, the actual behavior seems pretty intuitive to me.

Reply:
Now it's even more intuitive :)